File: python/n38_CountSay.py
# https://leetcode-cn.com/problems/count-and-say/
class Solution:
    def countAndSay(self, n: int) -> str:
        if n == 1:
            return '1'   
        if n == 2:
            return '11'
        last_count = self.countAndSay(n-1)
        this_count = ''
        num = 1
        for i in range(1, len(last_count)):
            if last_count[i] != last_count[i-1]:
                this_count = this_count + str(num) + last_count[i-1]
                num = 1
                continue
            else:
                num += 1
        this_count = this_count + str(num) + last_count[-1]      
        return this_count


if __name__ == "__main__":
    solu = Solution()
    print(solu.countAndSay(10))
    

File: cpp/n38_CountSay.cpp
// https://leetcode-cn.com/problems/count-and-say/
#include <iostream>
#include <string>
#include <vector>
using namespace std;
class Solution {
public:
    string countAndSay(int n) {
        if(n == 1){
            string say = "1";
            return say;
        }
        string last_say = countAndSay(n-1);
        string this_say = "";
        int num = 1;
        for(int i = 0; i < last_say.size(); ++i){
            if(i > 0 && last_say[i] != last_say[i-1]){
                this_say = this_say + to_string(num) + last_say[i-1];
                num = 1;
            }
            if(i > 0 && last_say[i] == last_say[i-1]) ++num;
        }
        this_say = this_say + to_string(num) + last_say[last_say.size()-1];
        return this_say;
    }
};

int main(){
    Solution solu;
    cout << solu.countAndSay(1) << endl;
    cout << solu.countAndSay(2) << endl;
    cout << solu.countAndSay(3) << endl;
    cout << solu.countAndSay(4) << endl;
    cout << solu.countAndSay(5) << endl;
}